src/parser/core/modules/ActionWindow/windows/ActionWindow.tsx

```diff
@@ -0,0 +1,203 @@
+	 */
+	protected onWindowStart(timestamp: number) {
+		this.history.getCurrentOrOpenNew(timestamp)
+	}
+	/**
+	 * Ends an existing window if one is open.
+	 * @param timestamp The timestamp at which the window ends.
+	 */
+	protected onWindowEnd(timestamp: number) {
+		this.history.closeCurrent(timestamp)
+	}
+	/**
+	 * Adds an action to the current window if one is open.
+	 * If no window is open, the event is ignored.
+	 * Implementing moudles MUST call removeDefaultActionHook before calling this method directly.
+	 * @param event The event to be added to the window.
+	 */
+	protected onWindowAction(event: Events['action']) {
+		this.history.doIfOpen(current => current.push(event))
+	}
+
+	/**
+	 * Removes the default event hook that captures all actions by the player.
+	 * Implementing modules should call this method if they have logic to only
+	 * include some actions in a window.
+	 * Implmenting modules MUST register their own hook that calls onWindowAction
+	 * after calling this method.
+	 */
+	protected removeDefaultActionHook() {
+		if (this.defaultEventHook != null) {
+			this.removeEventHook(this.defaultEventHook)
+			this.defaultEventHook = undefined
+		}
+	}
+	/**
+	 * Adjusts the default event hook to ignore certain actions.
+	 * Implementing modules MAY call this method if all casts of certain
+	 * actions should be ignored in a window.
+	 * If actions are only ignored in some conditions, this method is
+	 * not suitable, and you will need to register your own hook and callback
+	 * that only calls onWindowAction when the conditions are met.
+	 * Calling this method will override previous calls to trackOnlyActions.
+	 * @param actionsToIgnore The ids of the actions to ignore.
+	 */
+	protected ignoreActions(actionsToIgnore: number[]) {
+		this.removeDefaultActionHook()
+		this.defaultEventHook = this.addEventHook(
+			filter<Event>()
+				.source(this.parser.actor.id)
+				.action(noneOf(actionsToIgnore))
+				.type('action'),
+			this.onWindowAction)
+	}
+	/**
+	 * Adjusts the default event hook to only track certain actions.
+	 * Implementing modules MAY call this method if only some actions should
+	 * be tracked in a window.
+	 * If other actions should be tracked in some conditions, this method is
+	 * not suitable, and you will need to register your own hook and callback
+	 * that only calls onWindowAction when the conditions are met.
+	 * Calling this method will override previous calls to ignoreActions.
+	 * @param actionsToTrack The ids of the actions to track.
+	 */
+	protected trackOnlyActions(actionsToTrack: number[]) {
+		this.removeDefaultActionHook()
+		this.defaultEventHook = this.addEventHook(
+			filter<Event>()
+				.source(this.parser.actor.id)
+				.action(oneOf(actionsToTrack))
+				.type('action'),
+			this.onWindowAction)
+	}
+
+	override initialise() {
+		this.defaultEventHook = this.addEventHook(filter<Event>().source(this.parser.actor.id).type('action'), this.onWindowAction)
```

Comment:
Something definitely would have to change on the interface, since derived classes can set their own hooks.  I have an idea or two about how to do it, so I'll see what makes sense.

Reply:
Probably the easiest way is to replace the default hook property with something that accepts an Event filter object. That way you can add/remove the hook and pass the current filter object in unconditionally.

Now that I think about it more, that also simplifies the overriding story - you no longer have to remember to call a remove before adding your own, can just override the filter property and it inherently replaces the default.

Reply:
I get the idea here, but I'm not sure what type I need to declare for the variable that will hold the filter.  Based on the `addEventHook` method, my guess would be `Event['type'] | Partial<Event> | EventFilterPredicate<Event>`